File: Data Structures & Algorithms/binary-tree-diameter/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        dfs(ans, root);
        return ans;
    }
    int dfs(int &ans, TreeNode * root){
        if(!root) return 0;
        int left = dfs(ans, root->left);
        int right = dfs(ans, root->right);
        ans = max(left + right, ans);
        return 1+max(left, right);
    }
};

File: Data Structures & Algorithms/clone-graph/submission-1.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        curr = node
        mp = {}
        def dfs(curr):
            if not curr: return
            if curr in mp:
                return mp[curr]
            new = Node(curr.val)
            mp[curr] = new
            for nei in curr.neighbors:
                new.neighbors.append(dfs(nei))
            return mp[curr]
        return dfs(curr)
        